### PR TITLE
Allow reserved pages in toctree

### PR DESCRIPTION
**Pull request by avylove**
I was looking to add the general index to a toctree and it didn't seem like that was possible, so I created this patch to allow it.

Basically this allows the special document names (genindex, modindex, search) to be used as page names in a toctree either with their default titles or custom ones.

``` reStructuredText
.. toctree::

   normalpage
   genindex
   Modules <modindex>
   search
```

It seems I'm not the only one who was looking for this:

http://stackoverflow.com/questions/36235578/how-can-i-include-the-genindex-in-a-sphinx-toc
http://stackoverflow.com/questions/25243482/how-to-add-sphinx-generated-index-to-the-sidebar-when-using-read-the-docs-theme

### CODE REVIEW

File: doc/markup/toctree.rst [hunk 1]
```diff
@@ -181,6 +182,9 @@ tables of contents.  The ``toctree`` directive is the central element.
    .. versionchanged:: 1.3
       Added "caption" and "name" option.
 
+   .. versionchanged:: 1.5
```
Comment (Blendify) [inline doc/markup/toctree.rst]: Should be 4.5

**Comment by avylove** [comment]
Synced upstream changes and updated code as it was originally from 5.5 years ago. Also updated tests.

**Review by Blendify — APPROVED** [review]

File: doc/usage/restructuredtext/directives.rst [hunk 2]
```diff
@@ -220,6 +221,9 @@ tables of contents.  The ``toctree`` directive is the central element.
    .. versionchanged:: 1.3
       Added "caption" and "name" option.
 
+   .. versionchanged:: 4.5
```
Comment (AA-Turner) [inline doc/usage/restructuredtext/directives.rst:224]:
```suggestion
   .. versionchanged:: 5.1
```

If you re-target the PR to `5.x`

File: sphinx/directives/other.py [hunk 3]
```diff
@@ -124,7 +124,7 @@ def parse_content(self, toctree: addnodes.toctree) -> List[Node]:
                         break
                 # absolutize filenames
                 docname = docname_join(self.env.docname, docname)
-                if url_re.match(ref) or ref == 'self':
+                if url_re.match(ref) or ref in ('self', 'genindex', 'modindex', 'search'):
```
Comment (AA-Turner) [inline sphinx/directives/other.py:127]: I'd prefer if this 4-tuple was factored out somewhere as it is used more than once, but there aren't any great locations that come to mind.

File: doc/usage/restructuredtext/directives.rst [hunk 4]
```diff
@@ -150,7 +150,8 @@ tables of contents.  The ``toctree`` directive is the central element.
 
    The special entry name ``self`` stands for the document containing the
    toctree directive.  This is useful if you want to generate a "sitemap" from
-   the toctree.
+   the toctree. The special document names ``genindex``, ``modindex``, and
+   ``search`` can also be used.
```
Comment (AA-Turner) [inline doc/usage/restructuredtext/directives.rst:154]: Perhaps expand this, as the special names can also be used indirectly, as you illustrate in the test.

**Review by AA-Turner — CHANGES_REQUESTED** [review]
Please could you rebase on `5.x` and re-target the PR to that branch?

A

**Comment by AA-Turner** [comment]
Closed by #10673

**Pull request closed without merging**